### PR TITLE
Refactor tar truncator to support record based tar files.

### PR DESCRIPTION
**Pull request by Dorvaryn**
Tar files that are large are usually written as records, that means that we might get padded zeros in writing those chuncks.

That was breaking the truncation logic that was in place.

The new approach is more resilient as it will read by records of 20 blocking factor (default blocking factor of tar) and wait for the end of the stream to try and backtrack within the last read record to find the last block entry before the EOF marker. It will then copy the record up to that point.

This will work for all record size smaller or equal to the default record size of 20.
We could parametrize the blocking factor to make this working with an arbitrary blocking factor.

What do you think ?

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java [hunk 1]
```diff
@@ -17,11 +17,22 @@ public TarTruncator(DataWriter dataWriter) {
...
+            byte[] swappingRef;
+            int read;
+            int readLast = 0;
+
+            while ((read = readRecord(in, buffer)) > 0) {
+                newState = dataWriter.write(newState, buffer2, readLast);
+                swappingRef = buffer2;
+                buffer2 = buffer;
+                buffer = swappingRef;
+                readLast = read;
```
Comment (jonreeve) [inline library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java]: This is a bit hard to parse atm. Get it eventually, but some renaming would help a bit

File: library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java [hunk 2]
```diff
@@ -31,31 +42,30 @@ public TarTruncator(DataWriter dataWriter) {
...
 
-    private boolean isNotFullOfZeroes(byte[] buffer) {
-        for (byte b : buffer) {
-            if (b != BYTE_ZERO) {
-                return true;
-            }
+    private int truncateEOFMarker(byte[] buffer, int length) {
+        int position = length;
+
+        while (position >= Constants.TAR_BLOCK_SIZE && isFullOfZeroes(buffer, position - Constants.TAR_BLOCK_SIZE, position)) {
```
Comment (jonreeve) [inline library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java]: isFullOfZeroes - nice!

File: library/src/main/java/com/novoda/downloadmanager/lib/Constants.java [hunk 3]
```diff
@@ -108,6 +108,11 @@
     public static final int TAR_BLOCK_SIZE = 512;
 
     /**
+     * The size of a tar record
+     */
+    public static final int TAR_RECORD_SIZE = TAR_BLOCK_SIZE * 20;
```
Comment (ataulm) [inline library/src/main/java/com/novoda/downloadmanager/lib/Constants.java]: >The new approach is more resilient as it will read by records of 20 blocking factor (default blocking factor of tar)

You mentioned parameterising the 20 - yday I recommended doing it as part of this PR and leave 20 as the default. but now thinking on it, yagni; let's do it when someone raises the issue. The 20 can be extracted as a private constant though as `DEFAULT_TAR_BLOCKING_FACTOR` as it's not an arbitrary number.

File: library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java [hunk 4]
```diff
@@ -31,31 +42,30 @@ public TarTruncator(DataWriter dataWriter) {
...
+            if (buffer[i] != BYTE_ZERO) {
+                return false;
+            }
         }
-        return read;
+        return true;
     }
 
-    private int readBlock(InputStream fileInputStream, byte[] buffer) throws IOException {
+    private int readRecord(InputStream fileInputStream, byte[] buffer) throws IOException {
```
Comment (ataulm) [inline library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java]: could be static

File: library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java [hunk 5]
```diff
@@ -31,31 +42,30 @@ public TarTruncator(DataWriter dataWriter) {
         }
     }
 
-    private boolean isNotFullOfZeroes(byte[] buffer) {
-        for (byte b : buffer) {
-            if (b != BYTE_ZERO) {
-                return true;
-            }
+    private int truncateEOFMarker(byte[] buffer, int length) {
```
Comment (ataulm) [inline library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java]: `truncateEofMarker` - acronyms are treated as words

Comment (ataulm) [inline library/src/main/java/com/novoda/downloadmanager/lib/TarTruncator.java]: is this method more `findMarkerToTruncateEof()`. atm (truncate being a verb) it sounds like it will truncate the EOF marker and I don't think it is.

**Comment by ataulm** [comment]
the explanation you gave yesterday made sense, and this seems to line up with your explanation but I'm not confident enough reading it to assert that it lines up -> :+1:

**Comment by jonreeve** [comment]
LGTM :+1:
